File: Tutorials/Codes/Linear_Control_Course/Chapter27/Lesson5/Chapter27_Lesson5.py
import numpy as np

# Controller parameters from design
Kp = 3.2
Ki = 9.0

# Simulation parameters
dt = 0.001     # time step [s]
t_final = 8.0  # final time [s]
n_steps = int(t_final / dt) + 1

# Allocate arrays
t = np.linspace(0.0, t_final, n_steps)
y = np.zeros(n_steps)   # plant output
z = np.zeros(n_steps)   # integral state
u = np.zeros(n_steps)   # control signal
d = np.zeros(n_steps)   # disturbance

# Reference: unit step
def reference(time):
    return 1.0

# Disturbance: step applied at t >= 3 s
def disturbance(time):
    return 0.2 if time >= 3.0 else 0.0

for k in range(n_steps - 1):
    r_k = reference(t[k])
    d_k = disturbance(t[k])
    e_k = r_k - y[k]

    # PI control law
    u[k] = Kp * e_k + Ki * z[k]

    # Integral state derivative
    dz = e_k

    # Plant: y_dot = -y + u + d
    dy = -y[k] + u[k] + d_k

    # Euler integration
    z[k + 1] = z[k] + dt * dz
    y[k + 1] = y[k] + dt * dy

# Last control and disturbance
u[-1] = Kp * (reference(t[-1]) - y[-1]) + Ki * z[-1]
d[-1] = disturbance(t[-1])

# At this point, t, y, u, d contain the closed-loop trajectory.
# In a robotic joint controller, similar logic would be embedded
# in a 1 kHz control loop using python-control or roboticstoolbox.

File: Tutorials/Codes/Linear_Control_Course/Chapter27/Lesson5/Chapter27_Lesson5.java
public class PiTrackingSimulation {

    public static void main(String[] args) {
        double Kp = 3.2;
        double Ki = 9.0;

        double dt = 0.001;
        double tFinal = 8.0;
        int nSteps = (int) (tFinal / dt) + 1;

        double[] t = new double[nSteps];
        double[] y = new double[nSteps];
        double[] z = new double[nSteps];
        double[] u = new double[nSteps];
        double[] d = new double[nSteps];

        for (int k = 0; k < nSteps; ++k) {
            t[k] = k * dt;
        }

        for (int k = 0; k < nSteps - 1; ++k) {
            double r = 1.0; // unit step
            double dist = (t[k] >= 3.0) ? 0.2 : 0.0;
            double e = r - y[k];

            // PI controller
            u[k] = Kp * e + Ki * z[k];

            double dz = e;
            double dy = -y[k] + u[k] + dist;

            z[k + 1] = z[k] + dt * dz;
            y[k + 1] = y[k] + dt * dy;
            d[k] = dist;
        }
        // final sample
        double rEnd = 1.0;
        double distEnd = (t[nSteps - 1] >= 3.0) ? 0.2 : 0.0;
        double eEnd = rEnd - y[nSteps - 1];
        u[nSteps - 1] = Kp * eEnd + Ki * z[nSteps - 1];
        d[nSteps - 1] = distEnd;

        System.out.println("Final output y(T) = " + y[nSteps - 1]);

        // In a robotics Java stack, the same PI computation would run
        // in a periodic control task with joint states as feedback.
    }
}
